overlord/overlord.go

```diff
@@ -267,6 +267,11 @@ func (o *Overlord) Loop() {
 			// in case of errors engine logs them,
 			// continue to the next Ensure() try for now
 			o.stateEng.Ensure()
+			// without snaps on the system we may go into
+			// socket activation mode
+			if o.canGoSocketActivated() {
```

Comment:
The whole logic looks nice. There's just one thing that I cannot quite see yet: how do we make sure we stay alive for long enough to take a request?  Apparently it's just dying as soon as it gets started?  We need to keep in mind that the 5 minutes is a max, and not a min. So far we've been pretty liberal about calling ensure whenever we feel like it, which means we might easily create a situation where snapd won't take any calls.

Reply:
When we get a request the socket activation will wake snapd up and we can process the request. If this results in snaps getting installed snapd will not go into socket activation mode ever again. So that should be ok unless I miss something in the question.

Reply:
I think Gustavo's question is what make sure snapd stays up enough to accept the connection and process the request

Reply:
I can write a test for this, systemd will make sure connections are handled. I can create a test that waits for snapd to stop and once it stops fires a request (in a loop if desired) and see that its processed.

Reply:
Or is this about a race that if a request is already on its way we may shutdown the daemon and the request gets lost in that situation?

Reply:
it's not just a matter of systemd,  seems there's a race no between the connection being accepted up to creating a change, vs the first Ensure finishing and detecting there's no change

Reply:
As I understand it, systemd will make sure the connection is handled in terms of handing the socket off to snapd. The issue I'm wondering about is that snapd is committing suicide right when that's taking place. Do we have the proper logic to make sure the daemon won't terminate prematurely when that race takes place? One potential resilient solution for this is to refuse to die too fast, maybe?